### PR TITLE
Add isort config

### LINKED ISSUE #23048: Use isort for standardizing import formatting
@TomAugspurger kindly pointed out on this [PR](https://github.com/pandas-dev/pandas/pull/22913) that we could be using [isort](https://pypi.org/project/isort/) for formatting imports and have a standardized approach across the codebase.

I will draft up an initial config based on the rules used in [dask-ml](https://github.com/dask/dask-ml/blob/master/setup.cfg) and the comments from @jbrockmendel in the above PR. If anyone feels strongly about any of the rules please comment here.

My proposed plan for introducing is we define a config and I run this for a handful of files. ( 1st PR )

We then over the period of a few weeks we encourage developers to run isort over their PRs and we can iron out any issues in the config set up.

Then we introduce a check into our CI build `isort --recursive --check-only`. And perhaps use [this](https://pypi.org/project/flake8-isort/) so developers can just run flake8 on their code like before.

Any thoughts/concerns here?

### PR DESCRIPTION
**Pull request by alimcmaster1**
- [x] closes #23048
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
 
Confirming that CI fails when Imports are incorrectly formatted. See here:
https://travis-ci.org/pandas-dev/pandas/jobs/440557769

Sample output:

Doctests top-level reshaping functions DONE
Check import format using isort
ERROR: /home/travis/build/pandas-dev/pandas/pandas/core/series.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/core/indexes/range.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/io/packers.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/io/pytables.py Imports are incorrectly sorted.
Check import format using isort DONE
The command "ci/code_checks.sh" exited with 1.

### CODE REVIEW

**Review by datapythonista — COMMENTED** [review]
in general looks good, added some comments

**Comment by jreback** [comment]
pls rebase. let's lmit this first push to what you have done. don't add more files here. this will need many small PR's to avoid rebase issues generally.

File: setup.cfg [hunk 1]
```diff
@@ -76,3 +76,559 @@ exclude_lines =
...
+known_third_party=numpy,python-dateutil,pytz
+force_sort_within_sections=true
+forced_separate=pandas._libs,pandas.compat,pandas.util,pandas.core.dtypes,pandas.core,pandas.errors
+multi_line_output=5
+length_sort=True
+force_grid_wrap=0
+combine_as_imports=True
+skip=
+    pandas/lib.py,
+    pandas/tslib.py,
```
[the remark this answers is not in the document]
Comment (alimcmaster1) [reply setup.cfg]: Thanks - Yes it does i could run `isort -rc pandas` to sort the entire codebase but I will go for a more incremental approach suggested by @TomAugspurger [here](https://github.com/pandas-dev/pandas/issues/23048)

File: ci/travis-36.yaml [hunk 2]
```diff
@@ -14,6 +14,7 @@ dependencies:
   - geopandas
   - html5lib
   - ipython
+  - isort
```
[the remark this answers is not in the document]
Comment (alimcmaster1) [reply ci/travis-36.yaml]: Thanks have done added to `environment-dev.yaml`

**Comment by TomAugspurger** [comment]
Lint failure: https://travis-ci.org/pandas-dev/pandas/jobs/442319848#L2919

> Confirming that CI fails when Imports are incorrectly formatted.

Did CI fail without it? What did the output look like?

I think things were pretty quiet over in #23048, so just to confirm are any maintainers against doing this (@jorisvandenbossche @WillAyd @jschendel @gfyoung @mroeschke, ...)?

**Comment by alimcmaster1** [comment]
@TomAugspurger - thanks for taking a look below is an example CI failing for files with incorrectly formatted imports:

 https://travis-ci.org/pandas-dev/pandas/jobs/440557769#L2839

Sample output:

Doctests top-level reshaping functions DONE
Check import format using isort
ERROR: /home/travis/build/pandas-dev/pandas/pandas/core/series.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/core/indexes/range.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/io/packers.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/io/pytables.py Imports are incorrectly sorted.
Check import format using isort DONE
The command "ci/code_checks.sh" exited with 1.

**Comment by mroeschke** [comment]
I am on board in general, but will isort raise an error if imports are inlined in functions and not at the top of the file? I've seen them used here and there in the code base (maybe for a good reason).

Example:

https://github.com/pandas-dev/pandas/blob/913f71f74709e4d4814e50f21768b97ce2c6ef8d/pandas/core/tools/datetimes.py#L98

**Comment by TomAugspurger** [comment]
It only lints imports at the top of a file. I think it even stops linting if you do an import in a try/except.

________________________________

**Comment by jreback** [comment]
this looks fine. can you rebase, and ping on green.

**Comment by TomAugspurger** [comment]
@jreback green aside from the windows 3.6 failure. We should probably merge this before any others.

**Comment by alimcmaster1** [comment]
Thanks @TomAugspurger . We could also use [this](https://pypi.org/project/flake8-isort/), so any issues with imports will be highlighted when developers hopefully run `git diff upstream/master -u -- "*.py" | flake8 --diff` before committing. I can rebase this eve if required. ( this branch is 2 commits behind master )

**Comment by TomAugspurger** [comment]
Maybe merge mater to be safe.

Or you could spot check https://github.com/pandas-dev/pandas/pull/23190 and https://github.com/pandas-dev/pandas/pull/23193 to see if they touch files that are now being listed.

**Comment by jreback** [comment]
thanks @alimcmaster1

**Comment by jreback** [comment]
certainly take followups to remove the non-checked files, but do in smaller batches (doesn't have to be tiny), but not > 20 files

**Comment by alimcmaster1** [comment]
Thanks @jreback for sorting, apologies for lack of response from myself, unfortunately I cant push to public repo's on my work machine.. 

I will follow up with PRs for small batches of files. Let me know if there is anything I should be doing to document what's going on here :)

**Comment by datapythonista** [comment]
@alimcmaster1 if that makes sense, you can get the list of files to update, divide it in chunks, and create an issue for each chunk explaining well what needs to be done. This way you don't need to do all them yourself.

**Comment by alimcmaster1** [comment]
Thanks @datapythonista - I've started doing a few PRs myself and created an issue so others can hopefully help #23334.

**Comment by jorisvandenbossche** [comment]
@alimcmaster1 something else: would you like to add a small section to the contributing guideline explaining `isort` and its use? So we have something to point to if contributors on a PR get a red CI with "Imports are incorrectly sorted." (which sort order is expected? how to let isort do this automatically for you? ..)

**Comment by alimcmaster1** [comment]
Hey @jorisvandenbossche - sure I can do this, was thinking this might be handy! Thanks

**Comment by alimcmaster1** [comment]
@jorisvandenbossche thoughts on something like this #23364 ?

**Pull request merged**